docs/site/Accessing-http-request-response.md

```diff
@@ -0,0 +1,249 @@
+
+export class PingController {
+  constructor(
+    @inject(RestBindings.Http.CONTEXT) private requestCtx: RequestContext,
+  ) {}
+
+  ping(): Response {
+    const {request, response} = this.requestCtx;
+    this.response.status(200).send({
+      greeting: 'Hello from LoopBack',
+      date: new Date(),
+      url: this.req.url,
+      headers: Object.assign({}, this.req.headers),
+    });
+    // Return the HTTP response object so that LoopBack framework skips the
+    // generation of HTTP response
+    return response;
+  }
+}
+```
+
+### Use middleware to access HTTP request/response
+
+[REST middleware-based sequence](https://loopback.io/doc/en/lb4/REST-middleware-sequence.html)
+allows middleware to access HTTP request context as follows:
+
+```ts
+import {Middleware, MiddlewareContext} from '@loopback/rest';
+import {debugFactory} from 'debug';
+import {v1} from 'uuid';
+const trace = debugFactory('trace:request-response');
+
+/**
+ * An middleware for tracing HTTP requests/responses
+ * @param ctx - Context object
+ * @param next - Downstream middleware/handlers
+ */
+export const tracingMiddleware: Middleware = async (ctx, next) => {
+  setupRequestId(ctx);
+  const {request, response} = ctx;
+  try {
+    if (trace.enabled) {
+      const reqObj = {
+        method: request.method,
+        originalUrl: request.originalUrl,
+        headers: request.headers,
+        // Body is not available yet before `parseParams`
+      };
+      trace('Request: %s', reqObj);
+    }
+    const result = await next();
+    if (trace.enabled) {
+      const resObj = {
+        statusCode: response.statusCode,
+        headers: response.getHeaders(),
+      };
+      trace('Response: %s', resObj);
+    }
+    return result;
+  } catch (err) {
+    if (trace.enabled) {
+      trace('Error: %s', err);
+    }
+    throw err;
+  }
+};
+
+function setupRequestId(ctx: MiddlewareContext) {
+  let requestId = ctx.request?.get('X-Request-ID');
+  debug(
+    'RequestID for %s %s: %s',
+    ctx.request.method,
+    ctx.request.originalUrl,
+    requestId,
+  );
+  if (requestId == null) {
+    requestId = v1();
+    debug(
+      'A new RequestID is generated for %s %s: %s',
+      ctx.request.method,
+      ctx.request.originalUrl,
+      requestId,
+    );
+  }
+
+  // Bind `request.id` so that it is available for injection in downstream
+  // controllers/services
+  ctx.bind('request.id').to(requestId);
+}
+```
+
+## Using interceptors to access HTTP request/response
+
+[Interceptors](Interceptor.md) can access HTTP request context as follows:
+
+```ts
+import {Interceptor, InvocationContext} from '@loopback/core';
+import {RestBinding} from '@loopback/rest';
+
+/**
+ * An interceptor to access HTTP requests/responses
+ * @param ctx - Invocation Context object
+ * @param next - Downstream interceptors
+ */
+export const myInterceptor: Interceptor = async (ctx, next) => {
+  const reqCtx = await ctx.get(RestBinding.Http.CONTEXT);
+  // ...
+};
```

Comment:
I am not a big fan of using Service Locator pattern (`ctx.get`), I prefer to use Dependency Injection (`@inject`) instead. In my experience, code using Service Locators is more difficult to test in isolation (unit-test style), because the dependencies are not clear.

Is it possible to implement this interceptor as a class and use `@inject` for accessing the HTTP context?